README.md
Latest:
[![Hackage](https://img.shields.io/hackage/v/mercury-api.svg)](https://hackage.haskell.org/package/mercury-api)
Linux:
[![Build Status](https://travis-ci.org/ppelleti/hs-mercury-api.svg?branch=master)](https://travis-ci.org/ppelleti/hs-mercury-api)
Windows:
[![Build status](https://ci.appveyor.com/api/projects/status/aywuy9y05ow8wja2/branch/master?svg=true)](https://ci.appveyor.com/project/ppelleti/hs-mercury-api/branch/master)

## Description

This package is a Haskell binding to the [Mercury API][5] C API for
[ThingMagic][6] RFID readers.  It is especially geared toward
the [SparkFun Simultaneous RFID Reader][1], which uses ThingMagic's
[M6e Nano][7] module, but it should work with other ThingMagic
readers.  (Though currently, only support for serial readers is
compiled in.)  Most of the function and type names are the same as
their counterparts in the C API, with the `TMR_` prefix dropped.  For
more in-depth, language-independent documentation of Mercury API, see
[Mercury API Programmers Guide][2].

This package includes a copy of the Mercury API C library, so no
external libraries are necessary.  Several small bug fixes have been
applied to the included version of the library.  (I have submitted
these patches upstream, but I don't know if or when they will be
included in the official version.)  If you need to upgrade to a newer
version of Mercury API than the included one, see [UPGRADING.md][9].

The Haskell binding doesn't support background reads.  I recommend
that you just spawn a new Haskell thread and do foreground reads
instead.

Currently, only support for the serial reader is compiled in, but it
probably wouldn't be too hard to enable LLRP support.  (I don't have
any way to test LLRP, however, as the M6e Nano doesn't support it.)

On Mac OS X, be sure to use the serial device that starts with
`/dev/cu.`, not the serial device that starts with `/dev/tty.`.

On Windows, you can type `wmic path Win32_SerialPort` to list the
available COM ports.

Only some parameters and some tagops are currently supported in the
Haskell binding.  (There are a lot of them, and I only implemented the
ones I needed.)  If you need support for additional parameters or
tagops, please file an issue in GitHub and I will add them.

I have tested this package on Linux, Mac OS X, and Windows, using the
SparkFun board.

Here is a minimal example which reads tags at maximum power for 1
second, and then prints the results:

```haskell
{-# LANGUAGE OverloadedStrings #-}

import qualified Data.Text.IO as T
import qualified System.Hardware.MercuryApi as TMR
import qualified System.Hardware.MercuryApi.Params as TMR

main = do
  rdr <- TMR.create "tmr:///dev/ttyUSB0"
  TMR.paramSetTransportTimeout rdr 10000
  TMR.connect rdr
  TMR.paramSetBasics rdr TMR.REGION_NA2 2700 TMR.sparkFunAntennas
  tags <- TMR.read rdr 1000
  putStrLn $ "read " ++ show (length tags) ++ " tags"
  mapM_ T.putStrLn $ concatMap TMR.displayTagReadData tags
  TMR.destroy rdr
```

Additional examples are available in the `examples` directory.

## How to build

This package is mostly of interest to Haskell programmers, and if
you're a Haskell programmer, you already know how to build it with
`stack` or `cabal`.

However, some of the utilities like `tmr-firmware` might be of
interest to non-Haskellers, so here is how to build them if you don't
know anything about Haskell:

* Install [stack][11].

* Clone this repository.

* Run `stack --install-ghc install` at the top level of the repository.

* Add `~/.local/bin` to your `$PATH`.

Now the following utilities will be available in your `$PATH`:

```
tmr-firmware - write a new firmware image to reader

Usage: tmr-firmware [-u|--uri URI] [-t|--transport-listener] [FILENAME]
  If run without firmware file, just print current firmware version. Firmware
  file can be downloaded from:
  http://www.thingmagic.com/index.php/download-nano-firmware

Available options:
  -h,--help                Show this help text
  -u,--uri URI             Reader to connect to (default tmr:///dev/ttyUSB0)
  -t,--transport-listener  Print bytes sent on serial port


tmr-params - print parameters

Usage: tmr-params [-u|--uri URI] [-t|--transport-listener]

Available options:
  -h,--help                Show this help text
  -u,--uri URI             Reader to connect to (default tmr:///dev/ttyUSB0)
  -t,--transport-listener  Print bytes sent on serial port


tmr-read - read tags

Usage: tmr-read [-u|--uri URI] [-r|--region REGION] [-p|--power CENTI-DBM]
                [-t|--transport-listener] [-l|--long]

Available options:
  -h,--help                Show this help text
  -u,--uri URI             Reader to connect to (default tmr:///dev/ttyUSB0)
  -r,--region REGION       Regulatory region (default na2)
  -p,--power CENTI-DBM     Power level (0-2700, default 2300)
  -t,--transport-listener  Print bytes sent on serial port
  -l,--long                Print lots of information per tag


tmr-write - write a string to user data of tag

Usage: tmr-write [-u|--uri URI] [-r|--region REGION] [-p|--power CENTI-DBM]
                 [-t|--transport-listener] [-R|--rewrite] STRING

Available options:
  -h,--help                Show this help text
  -u,--uri URI             Reader to connect to (default tmr:///dev/ttyUSB0)
  -r,--region REGION       Regulatory region (default na2)
  -p,--power CENTI-DBM     Power level (0-2700, default 2300)
  -t,--transport-listener  Print bytes sent on serial port
  -R,--rewrite             Write to a tag even if written before
```

## Resources

Additional resources:

* [RFID Basics][8]
* [SparkFun Simultaneous RFID Reader hookup guide][3]
* [ThingMagic manuals and firmware][4]

Mercury API in other languages:

* C, Java, and C# - officially supported by [Mercury API][5]
* [Python binding][10] by Petr Gotthard

[1]: https://www.sparkfun.com/products/14066
[2]: http://www.thingmagic.com/images/Downloads/Docs/MercuryAPI_ProgrammerGuide_for_v1.27.3.pdf
[3]: https://learn.sparkfun.com/tutorials/simultaneous-rfid-tag-reader-hookup-guide
[4]: http://www.thingmagic.com/index.php/manuals-firmware
[5]: http://www.thingmagic.com/index.php/manuals-firmware#Mercury_API
[6]: http://www.thingmagic.com/
[7]: http://www.thingmagic.com/index.php/embedded-rfid-readers/thingmagic-nano-module
[8]: https://learn.sparkfun.com/tutorials/rfid-basics
[9]: https://github.com/ppelleti/hs-mercury-api/blob/master/UPGRADING.md
[10]: https://github.com/gotthardp/python-mercuryapi
[11]: https://docs.haskellstack.org/en/stable/README/#how-to-install


ChangeLog.md
<!-- -*- text -*- (prevent Emacs from formatting ChangeLog incorrectly) -->

# Revision history for mercury-api

## 0.1.0.2  -- 2019-01-09

* Avoid including `xlocale.h` on Linux.  (Fixes [#1][1] and [#2][2].)

[1]: https://github.com/ppelleti/hs-mercury-api/issues/1
[2]: https://github.com/ppelleti/hs-mercury-api/issues/2

## 0.1.0.1  -- 2017-06-11

* Fixed a crash that could occur when peeking the mask of a TagFilter.  (I was reading the length as a 32-bit number instead of a 16-bit number, which could cause the length to be impossibly long.)

## 0.1.0.0  -- 2017-06-05

* First version. Released on an unsuspecting world.


examples/ExampleUtil.hs
{-# LANGUAGE OverloadedStrings #-}

module ExampleUtil
  ( createAndConnect
  , createConnectAndParams
  , optUri
  , optRegion
  , optPower
  , optListen
  ) where

import Control.Exception ( throw, try )
import Control.Monad ( when, void, forM_ )
import Data.Int ( Int32 )
import Data.Monoid ( (<>) )
import qualified Data.Text as T ( replicate, pack, length )
import qualified Data.Text.IO as T ( putStrLn )
import Options.Applicative
    ( Parser,
      value,
      switch,
      strOption,
      short,
      option,
      metavar,
      long,
      help,
      auto )
import qualified System.Hardware.MercuryApi as TMR
import qualified System.Hardware.MercuryApi.Params as TMR
import System.Exit ( exitFailure )
import System.Info ( os )
import System.IO ( stdout )

printRegionsAndFail :: TMR.Reader -> IO a
printRegionsAndFail rdr = do
  rgns <- TMR.paramGetRegionSupportedRegions rdr
  T.putStrLn "Region must be one of:"
  forM_ rgns $ \rgn -> do
    let dr = TMR.displayRegion rgn
        nSpaces = 6 - T.length dr
        spaces = T.replicate nSpaces " "
    T.putStrLn $ "  " <> dr <> spaces <> TMR.displayRegionDescription rgn
  exitFailure

parseRegionOrFail :: TMR.Reader -> String -> IO TMR.Region
parseRegionOrFail rdr s =
  case TMR.parseRegion (T.pack s) of
    Nothing -> printRegionsAndFail rdr
    Just rgn -> return rgn

printPowerAndFail :: TMR.Reader -> IO a
printPowerAndFail rdr = do
  lo <- TMR.paramGetRadioPowerMin rdr
  hi <- TMR.paramGetRadioPowerMax rdr
  putStrLn $ "Power must be between " ++ show lo ++ " and " ++ show hi
  exitFailure

handleParamError :: TMR.Reader -> Either TMR.MercuryException () -> IO ()
handleParamError _ (Right _) = return ()
handleParamError rdr (Left err) = hpe (TMR.meStatus err)
  where hpe TMR.ERROR_INVALID_REGION = printRegionsAndFail rdr
        hpe TMR.ERROR_MSG_POWER_TOO_HIGH = printPowerAndFail rdr
        hpe TMR.ERROR_MSG_POWER_TOO_LOW = printPowerAndFail rdr
        hpe _ = throw err

createAndConnect :: String -> Bool -> IO TMR.Reader
createAndConnect uri listen = do
  rdr <- TMR.create $ T.pack uri
  when (listen) $ do
    listener <- TMR.opcodeListener stdout
    void $ TMR.addTransportListener rdr listener
  TMR.paramSetTransportTimeout rdr 10000
  TMR.connect rdr
  return rdr

createConnectAndParams :: String -> Bool -> String -> Int32 -> IO TMR.Reader
createConnectAndParams uri listen region power = do
  rdr <- createAndConnect uri listen
  rgn <- parseRegionOrFail rdr region
  eth <- try $ TMR.paramSetBasics rdr rgn power TMR.sparkFunAntennas
  handleParamError rdr eth
  TMR.paramSetTagReadDataRecordHighestRssi rdr True
  return rdr

defUri :: String
defUri = case os of
           "darwin" -> "tmr:///dev/cu.SLAB_USBtoUART"
           "mingw32" -> "tmr:///COM4"
           _ -> "tmr:///dev/ttyUSB0"

defRegion :: String
defRegion = "na2"

defPower :: Int32
defPower = 2300

optUri :: Parser String
optUri = strOption (long "uri" <>
                    short 'u' <>
                    metavar "URI" <>
                    help ("Reader to connect to (default " ++ defUri ++ ")") <>
                    value defUri)

optRegion :: Parser String
optRegion = strOption (long "region" <>
                       short 'r' <>
                       metavar "REGION" <>
                       help ("Regulatory region (default " ++ defRegion ++ ")") <>
                       value defRegion)

optPower :: Parser Int32
optPower = option auto (long "power" <>
                        short 'p' <>
                        metavar "CENTI-DBM" <>
                        help ("Power level (0-2700, default " ++ show defPower ++ ")") <>
                        value defPower)

optListen :: Parser Bool
optListen = switch (long "transport-listener" <>
                    short 't' <>
                    help "Print bytes sent on serial port")
